packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -25,36 +54,470 @@ export class LoaderContainerTracker {
+     */
+    private getPendingClients(containersToApply: IContainer[]) {
+        // All the clientId we track should be a superset of the quorum, otherwise, we are missing
+        // leave messages
+        const openedDocuments = Array.from(this.containers.keys()).filter((c) => !c.closed);
+        const openedClientId = openedDocuments.map((container) => (container as Container).clientId);
+
+        const pendingClients: [IContainer, Set<string>][] = [];
+        containersToApply.forEach((container) => {
+            const pendingClientId = new Set<string>();
+            const quorum = container.getQuorum();
+            quorum.getMembers().forEach((client, clientId) => {
+                // ignore summarizer
+                if (!client.client.details.capabilities.interactive) { return; }
+                if (!openedClientId.includes(clientId)) {
+                    pendingClientId.add(clientId);
+                }
             });
+
+            if (pendingClientId.size !== 0) {
+                pendingClients.push([container, pendingClientId]);
+            }
+        });
+        return pendingClients;
+    }
+
+    /**
+     * Utility to check synchronization based on sequence number
+     * See ensureSynchronized for more detail
+     *
+     * @param containersToApply - the set of containers to check
+     */
+    private isSequenceNumberSynchronized(containersToApply: IContainer[]) {
+        // clientSequenceNumber check detects ops in flight, both on the wire and in the outbound queue
+        // We need both client sequence number and isDirty check because:
+        // - Currently isDirty flag ignores ops for task scheduler, so we need the client sequence number check
+        // - But isDirty flags include ops during forceReadonly and disconnected, because we don't submit
+        //   the ops in the first place, clientSequenceNumber is not assigned
+
+        const isClientSequenceNumberSynchronized = containersToApply.every((container) => {
+            if (container.deltaManager.readOnlyInfo.readonly === true) {
+                // Ignore readonly container. the clientSeqNum and clientSeqNumObserved might be out of sync
+                // because we transition to readonly when outbound is not empty or the in transit op got lost
+                return true;
+            }
+            // Note that in read only mode, the op won't be submitted
+            const deltaManager = (container.deltaManager as any);
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const { trailingNoOps } = this.containers.get(container)!;
+            return deltaManager.clientSequenceNumber ===
+                (deltaManager.clientSequenceNumberObserved as number) + trailingNoOps;
+        });
+
+        if (!isClientSequenceNumberSynchronized) {
+            return false;
         }
+
+        const minSeqNum = containersToApply[0].deltaManager.minimumSequenceNumber;
```

Comment:
it's a bit weird we need to special case the quorum, won't other consensus structures have the same issue? i'd rather make the test wait if the side effect it cares about, approve or reject in this case

Reply:
All the consensus data structure only need the round trip.   The quorum proposal is unique in that  it allow other client to reject, thus needs to wait until minSeqNumber to catch up.

It is debatable that whether we should have the test wait for proposal accepted or not and just let op resume in the mean time.   But I do see a benefit to basically ensure the system has completely "settled".